native/dlaqr2.go

```diff
@@ -0,0 +1,382 @@
+	case nw < 0 || kbot-ktop+1 < nw:
+		panic("lapack: invalid value of nw")
+	}
+	if wantz {
+		checkMatrix(n, n, z, ldz)
+		switch {
+		case iloz < 0 || ktop < iloz:
+			panic("lapack: invalid value of iloz")
+		case ihiz < kbot || n <= ihiz:
+			panic("lapack: invalid value of ihiz")
+		}
+	}
+	checkMatrix(nw, nw, v, ldv)
+	checkMatrix(nw, nw, t, ldt)
+	checkMatrix(nw, nw, wv, ldwv)
+	switch {
+	case nh > nw:
+		panic("lapack: invalid value of nh")
+	case nv > nw:
+		panic("lapack: invalid value of nv")
+	case len(sr) != kbot+1:
+		panic("lapack: bad length of sr")
+	case len(si) != kbot+1:
+		panic("lapack: bad length of si")
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork != -1 && lwork < max(1, 2*nw):
+		panic(badWork)
+	}
+
+	// LAPACK code does not enforce the documented behavior
+	//  nw <= kbot-ktop+1
+	// but we do (we panic above).
+	jw := nw
+	lwkopt := max(1, 2*nw)
+	if jw > 2 {
+		// Workspace query call to Dgehrd.
+		impl.Dgehrd(jw, 0, jw-2, t, ldt, nil, work, -1)
+		lwk1 := int(work[0])
+		// Workspace query call to Dormhr.
+		impl.Dormhr(blas.Right, blas.NoTrans, jw, jw, 0, jw-2, t, ldt, work, v, ldv, work, -1)
+		lwk2 := int(work[0])
+		// Optimal workspace.
+		lwkopt = jw + max(lwk1, lwk2)
+	}
+
+	// Quick return in case of workspace query.
+	if lwork == -1 {
+		work[0] = float64(lwkopt)
+		return 0, 0
+	}
+
+	if nw == 0 {
+		return
+	}
+
+	// Machine constants.
+	ulp := dlamchP
+	smlnum := float64(n) / ulp * dlamchS
+
+	// Setup deflation window.
+	var s float64
+	kwtop := kbot - jw + 1
+	if kwtop != ktop {
+		s = h[kwtop*ldh+kwtop-1]
+	}
+	if kwtop == kbot {
+		// 1×1 deflation window.
+		sr[kwtop] = h[kwtop*ldh+kwtop]
+		si[kwtop] = 0
+		ns = 1
+		nd = 0
+		if math.Abs(s) <= math.Max(smlnum, ulp*math.Abs(h[kwtop*ldh+kwtop])) {
+			ns = 0
+			nd = 1
+			if kwtop > ktop {
+				h[kwtop*ldh+kwtop-1] = 0
+			}
+		}
+		work[0] = 1
+		return ns, nd
+	}
+
+	// Convert to spike-triangular form. In case of a rare QR failure, this
+	// routine continues to do aggressive early deflation using that part of
+	// the deflation window that converged using infqr here and there to
+	// keep track.
+	impl.Dlacpy(blas.Upper, jw, jw, h[kwtop*ldh+kwtop:], ldh, t, ldt)
+	bi := blas64.Implementation()
+	bi.Dcopy(jw-1, h[(kwtop+1)*ldh+kwtop:], ldh+1, t[ldt:], ldt+1)
+	impl.Dlaset(blas.All, jw, jw, 0, 1, v, ldv)
+	infqr := impl.Dlahqr(true, true, jw, 0, jw-1, t, ldt, sr[kwtop:], si[kwtop:], 0, jw-1, v, ldv)
+	// Note that ilo == 0 which conveniently coincides with the success
+	// value of infqr, that is, infqr as an index always points to the first
+	// converged eigenvalue.
+
+	// Dtrexc needs a clean margin near the diagonal.
+	for j := 0; j < jw-3; j++ {
+		t[(j+2)*ldt+j] = 0
+		t[(j+3)*ldt+j] = 0
+	}
+	if jw >= 3 {
+		t[(jw-1)*ldt+jw-3] = 0
+	}
+
+	ns = jw
+	ilst := infqr
+	// Deflation detection loop.
+	for ilst < ns {
+		var bulge bool
+		if ns >= 2 {
+			bulge = t[(ns-1)*ldt+ns-2] != 0
+		}
+		if !bulge {
```

Comment:
The code duplication here is troubling, but I don't see a way to factor it cleanly. I guess that's why it's like this.